### PR TITLE
[WIP] adding coveralls coverage support

### PR DESCRIPTION
**Pull request by trws**
Need the pull request up so travis will grab it.

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
This is awesome! Thanks @trws

**Comment by davidbeckingsale** [comment]
Is there a reason you dropped gcc-5 and clang 3.6 from the matrix?

**Comment by trws** [comment]
Mainly build time, they can come back if you think they're useful, but gcc 5 was never really stable, and clang-3.6 seems off the back end of the support curve for the most part.

**Comment by davidbeckingsale** [comment]
I don't mind them going, just checking it was intentional.

**Comment by davidbeckingsale** [comment]
This is pretty cool!

**Comment by trws** [comment]
Hoping to get both coveralls and codecov on here so we can see the difference.  The earlier coveralls thing wasn't working right.

**Comment by trws** [comment]
There it is.  Interesting how different those coverage numbers are...

**Comment by trws** [comment]
Ah, needs another exclude.

**Comment by trws** [comment]
This branch now has support for two free online coverage tools, coveralls and codecov.  We probably only want one in our standard testing suite, so have a look and see what you think of each of them. @rhornung67, @davidbeckingsale, @DavidPoliakoff

**Comment by DavidPoliakoff** [comment]
My vote is for codecov, I like the way it views the build quite a bit better as a way of analyzing what the status of the repo is. That said, I haven't seen how it handles "these are the changes you made, and how they impacted coverage"

**Comment by trws** [comment]
Right. I suppose we could merge it like this and decide later. Downside is we'll get double notifications for a while...

Sent from my iPhone

On Jul 27, 2016, at 7:32 PM, David Poliakoff <notifications@github.com<mailto:notifications@github.com>> wrote:


My vote is for codecov, I like the way it views the build quite a bit better as a way of analyzing what the status of the repo is. That said, I haven't seen how it handles "these are the changes you made, and how they impacted coverage"

-

**Comment by DavidPoliakoff** [comment]
Sounds wise to me. I do want to see us propose a branch that adds a bunch of stuff and see which most roundly calls us out for not having tests associated with the new feature, and to push a branch with new tests and see how the coverage improvement is reported. After seeing those I should be ready to make a decision

:+1:

**Comment by rhornung67** [comment]
👍 
@trws thanks for setting these up and both look reasonable. Still trying to understand what some of these things are telling me. For example, do you know why some source lines are marked as covered/not covered and adjacent lines are not marked. This appears to be the case with wither tool.  Also, why I think the coveralls "sunburst" figure looks nice, it's not clear to me what it shows.

I believe the main coverage deficiencies we are observing are due to only hitting the CilkPlus code with the Intel compiler, not exercising features like copy construction in our simple tests, etc.

I think tools like these will be especially useful for the performance suite and proxy app repo. Is there a way to generate coverage numbers for different executables -- for example, LULESH vs. Kripke?

**Comment by DavidPoliakoff** [comment]
The sunburst has the root of the repo at the center, with subdirectories coming off of it. The size of a segment is proportional to the number of statements in it, the green/red is level of coverage. So if I see a big red segment, I might click that, and at the bottom of the sunburst you'll see what that was representing (and it'll reroot on that segment of the project).

It's the kind of visualization where only people who really like them know what the heck is going on, but as somebody who likes them, it provides a hierarchy and two senses of importance ("size" of section and coverage) all in one picture, it's kinda neat if you play with it.

**Comment by trws** [comment]
Sure, we can get numbers like these for just about anything, it just means setting up extra tests or doing it manually with gcov.

As to adjacent lines showing covered versus not, I'd have to look at the examples but sometimes it's that a line isn't considered for coverage because it isn't executable, or because of a conditional or I limit or what have you. I forgot to compile with -O0, so it could be the inlining bit.

Sent from my iPhone

On Jul 28, 2016, at 9:14 AM, rhornung67 <notifications@github.com<mailto:notifications@github.com>> wrote:


?
@trws<https://github.com/trws> thanks for setting these up and both look reasonable. Still trying to understand what some of these things are telling me. For example, do you know why some source lines are marked as covered/not covered and adjacent lines are not marked. This appears to be the case with wither tool. Also, why I think the coveralls "sunburst" figure looks nice, it's not clear to me what it shows.

I believe the main coverage deficiencies we are observing are due to only hitting the CilkPlus code with the Intel compiler, not exercising features like copy construction in our simple tests, etc.

I think tools like these will be especially useful for the performance suite and proxy app repo. Is there a way to generate coverage numbers for different executables -- for example, LULESH vs. Kripke?

-

**Comment by keasler** [comment]
I'll take a guess as to why some lines show coverage and some don't.   There is insufficient DWARF information.  When we got the response email from VTune about perceived deficiencies, they said something about the DWARF info not always containing complete information.  Since both tools show the same result, I'm guessing this is likely the problem.   That said, seeing the code segment in question might make me guess differently.

**Comment by rhornung67** [comment]
@DavidPoliakoff I understand what you're saying about the sunburst thing. The thing I don't like about it is that I don't know what I'm clicking on until I click on it (i.e., the name of the file, directory, etc. shows up at the bottom). Then, there doesn't appear to be a way to back my selection. I have to reload the page in my browser and start over.

Is there an option to expose some text labels?

**Comment by rhornung67** [comment]
@keasler and @trws a couple of examples that show what I was asking about can be seen in the Coveralls report for IndexSet.cxx.
1) Look at the swap method that starts at line 102
2) Compare the lines where setIcount is called in push_back_private (line 397) vs. push_front_private (lines 412, and 419).

**Comment by DavidPoliakoff** [comment]
@rhornung67 : that's actually a surprisingly complicated question. At the center, the labels would be shorter (/src , /include), as you go out they'd get longer (/src/david_doing_stupid_things/ocaml_reimplementation) but you'd have less screen space in which to render them.

Challenge aside: yeah, they should have that.

**Comment by trws** [comment]
That definitely looks like the optimization issue, which causes many of 
the missing dwarf problems.  The missing line at 404 is just that the 
segment type is never invalid, so it never tests that code path, but the 
setIcount lines being missing almost has to be optimizations talking.

**Comment by keasler** [comment]
@rhornung67   Ok, yeah.  That's just weird.  It touches it in one case, but not the other.  I guess we have to look at all the call sites.

**Comment by jonesholger** [comment]
@DavidPoliakoff  I agree with your comment 

> That said, I haven't seen how it handles "these are the changes you made, and how they impacted coverage"

I'm personally interested in also putting in some tests to exercise more of /src and watching it turn from auburn to green in the sunburst and in the line by line coloring with the two coverage approaches. Is this branch officially a playground

**Comment by trws** [comment]
If anyone wants to play with this functionality, I would recommend fetching it and making a new branch on here for whatever you want to test so we can see how it works without having to back it out to be able to merge.

**Comment by davidbeckingsale** [comment]
Shall we go with coveralls for coverage and get this merged?

**Comment by trws** [comment]
This PR actually does both coveralls and codecov, it would need a line removed to bring it down to one service.

File: cmake/SetupPackages.cmake [hunk 1]
```diff
@@ -94,8 +94,6 @@ if (RAJA_ENABLE_TESTS)
       CMAKE_ARGS                
           -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
           -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
-          -DCMAKE_C_FLAGS=${CMAKE_C_FLAGS}
```
Comment (davidbeckingsale) [inline cmake/SetupPackages.cmake]: I feel like this was here for a reason, but I'm assuming it means coverage info gets generated for GoogleTest?

Comment (trws) [inline cmake/SetupPackages.cmake]: Poking at that might be a good idea.  I had something in there to ignore some of that, but I'm not sure I got it all.

**Comment by rhornung67** [comment]
When we discussed this at a meeting a while back, we agreed that both report the same info (since they’re based on gcov) but that codecov was easier to navigate and understand.  David P. and Tom S. were not at that meeting, so they should weigh in if they have a strong opinion either way.

**Comment by trws** [comment]
I'm good with either really.  The codecov website is nicer but slightly buggier, their collection script is a great deal easier to use as well, so I'm fine with that.




On October 27, 2016 at 2:05:26 PM PDT, rhornung67 <notifications@github.com> wrote:
When we discussed this at a meeting a while back, we agreed that both report the same info (since they're based on gcov) but that codecov was easier to navigate and understand. David P. and Tom S. were not at that meeting, so they should weigh in if they have a strong opinion either way.

**Comment by davidbeckingsale** [comment]
I've dropped the coveralls stuff, I think... 😄

**Comment by davidbeckingsale** [comment]
@trws was the `secure` part of `.travis.yml`  for CodeCov, Coveralls, or both?

**Comment by trws** [comment]
I think it was coveralls only... would have to look back at the logs of 
the original commits to be sure, but I think codecov just accepts the 
Travis OAuth identity.

**Comment by davidbeckingsale** [comment]
I've pulled out the `secure` stuff, and hopefully this will then be good to go. One thing - did we want to use the reduced compiler matrix, or stick with what we used to use?

**Pull request merged**